File: tools/sdks/compose/src/main/kotlin/io/thunder/compose/components/presentation/organization/CreateOrganization.kt
package io.thunder.compose.components.presentation.organization

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.BasicText
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import io.thunder.android.Organization
import io.thunder.compose.LocalThunder
import io.thunder.compose.components.actions.BaseThunderSignUpButton
import kotlinx.coroutines.launch

/** Form to create a new organization (spec §8.4 Presentation). */
@Composable
fun ThunderCreateOrganization(
    modifier: Modifier = Modifier,
    onCreated: ((Organization) -> Unit)? = null,
    onError: ((String) -> Unit)? = null,
) {
    val state = LocalThunder.current
    val i18n = state.i18n
    BaseThunderCreateOrganization(modifier = modifier, onCreated = onCreated, onError = onError) { name, handle, isLoading, error, create ->
        Column(modifier = Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
            BasicText(i18n.resolve("createOrganization.title"))
            error?.let { BasicText(it) }
            BasicTextField(
                value = name.value,
                onValueChange = { name.value = it },
                modifier = Modifier.fillMaxWidth().defaultMinSize(minHeight = 44.dp),
            )
            BasicTextField(
                value = handle.value,
                onValueChange = { handle.value = it },
                modifier = Modifier.fillMaxWidth().defaultMinSize(minHeight = 44.dp),
            )
            BaseThunderSignUpButton(label = i18n.resolve("createOrganization.submit"), modifier = Modifier.fillMaxWidth()) {
                if (!isLoading) create()
            }
        }
    }
}

/** Unstyled base variant (spec §8.3). */
@Composable
fun BaseThunderCreateOrganization(
    modifier: Modifier = Modifier,
    onCreated: ((Organization) -> Unit)? = null,
    onError: ((String) -> Unit)? = null,
    content: @Composable (name: MutableState<String>, handle: MutableState<String>, isLoading: Boolean, error: String?, create: () -> Unit) -> Unit,
) {
    val state = LocalThunder.current
    val scope = rememberCoroutineScope()
    val name = remember { mutableStateOf("") }
    val handle = remember { mutableStateOf("") }
    var isLoading by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    val create = {
        if (name.value.isNotBlank()) {
            scope.launch {
                isLoading = true; error = null
                try {
                    val org = state.client.createOrganization(
                        name = name.value,
                        handle = handle.value.takeIf { it.isNotBlank() },
                    )
                    isLoading = false
                    onCreated?.invoke(org)
                } catch (e: Exception) {
                    error = e.message
                    isLoading = false
                    onError?.invoke(e.message ?: "Failed to create organization")
                }
            }
        }
    }

    Box(modifier = modifier) { content(name, handle, isLoading, error, create) }
}

File: tools/sdks/swiftui/Sources/ThunderSwiftUI/components/presentation/organization/CreateOrganization.swift
import SwiftUI
import Thunder

/// Form to create a new organization (spec §8.4 Presentation).
public struct ThunderCreateOrganization: View {
    @EnvironmentObject private var i18n: ThunderI18n
    public let onCreated: ((Organization) -> Void)?
    public let onError: ((String) -> Void)?

    public init(onCreated: ((Organization) -> Void)? = nil, onError: ((String) -> Void)? = nil) {
        self.onCreated = onCreated
        self.onError = onError
    }

    public var body: some View {
        BaseThunderCreateOrganization(onCreated: onCreated, onError: onError) { name, handle, isLoading, error, create in
            VStack(alignment: .leading, spacing: 12) {
                Text(i18n.resolve("createOrganization.title"))
                    .accessibilityAddTraits(.isHeader)
                if let error { Text(error).foregroundColor(.red) }
                TextField(i18n.resolve("createOrganization.name"), text: name)
                    .accessibilityLabel(i18n.resolve("createOrganization.name"))
                    .frame(minHeight: 44)
                TextField(i18n.resolve("createOrganization.handle"), text: handle)
                    .accessibilityLabel(i18n.resolve("createOrganization.handle"))
                    .frame(minHeight: 44)
                Button(i18n.resolve("createOrganization.submit")) { create() }
                    .disabled(isLoading)
                    .frame(minWidth: 44, minHeight: 44)
            }
            .padding()
        }
    }
}

/// Unstyled base variant (spec §8.3).
public struct BaseThunderCreateOrganization<Content: View>: View {
    @EnvironmentObject private var state: ThunderState
    public let onCreated: ((Organization) -> Void)?
    public let onError: ((String) -> Void)?
    public let content: (Binding<String>, Binding<String>, Bool, String?, () -> Void) -> Content

    @State private var name = ""
    @State private var handle = ""
    @State private var isLoading = false
    @State private var error: String?

    public init(
        onCreated: ((Organization) -> Void)? = nil,
        onError: ((String) -> Void)? = nil,
        @ViewBuilder content: @escaping (Binding<String>, Binding<String>, Bool, String?, () -> Void) -> Content
    ) {
        self.onCreated = onCreated
        self.onError = onError
        self.content = content
    }

    public var body: some View {
        content($name, $handle, isLoading, error, create)
    }

    private func create() {
        guard !name.isEmpty else { return }
        isLoading = true; error = nil
        Task {
            do {
                let org = try await state.client.createOrganization(
                    name: name,
                    handle: handle.isEmpty ? nil : handle
                )
                isLoading = false
                onCreated?(org)
            } catch {
                self.error = error.localizedDescription
                isLoading = false
                onError?(error.localizedDescription)
            }
        }
    }
}
